File: src/main/java/com/github/solairerove/algs4/leprosorium/arrays/BinarySearch.java
package com.github.solairerove.algs4.leprosorium.arrays;

import java.util.Arrays;

public class BinarySearch {
    public static void main(String[] args) {
        int[] a = new int[]{1, 1, 2, 3, 5, 8, 11, 15, 33};

        System.out.printf("For array: %s\n", Arrays.toString(a));
        System.out.printf("Index of element %s is %s\n", 8, rank(8, a)); // 5
        System.out.printf("Index of element %s is %s\n", 8, ownRank(8, a));
    }

    public static int rank(int key, int[] a) {
        return rank(key, a, 0, a.length - 1);
    }

    public static int rank(int key, int[] a, int lo, int hi) {
        if (lo > hi) {
            return -1;
        }

        int mid = lo + (hi - lo) / 2;

        if (key < a[mid]) {
            return rank(key, a, lo, mid - 1);
        } else if (key > a[mid]) {
            return rank(key, a, mid + 1, hi);
        } else {
            return mid;
        }
    }

    public static int ownRank(int key, int[] a) {
        int low = 0;
        int high = a.length - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (key < a[mid]) {
                high = mid - 1;
            } else if (key > a[mid]) {
                low = mid + 1;
            } else {
                return mid;
            }
        }

        return -1;
    }
}

File: src/main/kotlin/com/github/solairerove/algs4/leprosorium/searching/BinarySearch.kt
package com.github.solairerove.algs4.leprosorium.searching

fun main() {
    val items = listOf(5, 6, 4, 3, 10, 9, 5, 6, 7)
    print("items: $items \n")

    val sorted = items.sorted()
    print("items: $items \n")
    print("trying to find 10, index: ${binarySearch(sorted, 10)} - ${rank(sorted, 10)} \n") // 8
    print("trying to find 10, index: ${binarySearch(sorted, 6)} - ${rank(sorted, 6)} \n") // 4
    print("trying to find 10, index: ${binarySearch(sorted, 0)} - ${rank(sorted, 0)} \n") // -1
}

// O(log(n)) time | O(1) space
private fun binarySearch(items: List<Int>, target: Int): Int {
    var low = 0
    var high = items.size - 1

    while (low <= high) {
        val mid = low + (high - low) / 2

        when {
            items[mid] < target -> low = mid + 1
            items[mid] > target -> high = mid - 1
            else -> return mid
        }
    }

    return -1
}

// O(log(n)) time | O(log(n)) space cause of stack call
private fun rank(items: List<Int>, item: Int): Int {
    return rank(items, item, 0, items.size)
}

private fun rank(items: List<Int>, item: Int, low: Int, high: Int): Int {
    if (low > high) return -1

    val mid = (low + high) / 2
    val guess = items[mid]

    return when {
        guess == item -> mid
        guess > item -> rank(items, item, low, mid - 1)
        else -> rank(items, item, mid + 1, high)
    }
}
